x/mongo/driver/topology/server_test.go

```diff
@@ -179,59 +179,145 @@ func TestServer(t *testing.T) {
-				Name:            "",
 				Code:            10107,
-				Message:         "not master",
-				Details:         []byte{},
-				Labels:          []string{},
-				TopologyVersion: nil,
+				TopologyVersion: oldTV,
 			},
 		}
-		s.ProcessError(wce, initConnection{})
-
-		// should set ServerDescription to Unknown
-		resultDesc := s.Description()
-		require.Equal(t, resultDesc.Kind, (description.ServerKind)(description.Unknown))
-		require.Equal(t, resultDesc.LastError, wce)
-
-		// pool should be drained
-		if s.pool.generation < 1 {
-			t.Errorf("Expected pool to be drained once from a write concern error. got %d; want %d", s.pool.generation, 1)
+		newNotMasterWCError := driver.WriteCommandError{
+			WriteConcernError: &driver.WriteConcernError{
+				Code: 10107,
+			},
+		}
+		newShutdownWCError := driver.WriteCommandError{
+			WriteConcernError: &driver.WriteConcernError{
+				Code: 11600,
+			},
+		}
+		nonStateChangeError := driver.Error{
+			Code: 1,
+		}
+		networkTimeoutError := driver.Error{
+			Labels: []string{driver.NetworkError},
+			Wrapped: ConnectionError{
+				// Use a net.Error implementation that can return true from its Timeout() function.
+				Wrapped: &net.DNSError{
+					IsTimeout: true,
+				},
+			},
+		}
+		contextCanceledError := driver.Error{
+			Labels: []string{driver.NetworkError},
+			Wrapped: ConnectionError{
+				Wrapped: context.Canceled,
+			},
+		}
+		nonTimeoutNetworkError := driver.Error{
+			Labels: []string{driver.NetworkError},
+			Wrapped: ConnectionError{
+				// Use a net.Error implementation that always returns false from its Timeout() function.
+				Wrapped: &net.AddrError{},
+			},
 		}
-	})
-	t.Run("no WriteConcernError", func(t *testing.T) {
-		s, err := NewServer(address.Address("localhost"), primitive.NewObjectID())
-		require.NoError(t, err)
-
-		var desc *description.Server
-		descript := s.Description()
-		desc = &descript
-		require.Nil(t, desc.LastError)
-		s.connectionstate = connected
-		s.pool.connected = connected
-
-		wce := driver.WriteConcernError{}
-		s.ProcessError(&wce, initConnection{})
 
-		// should not be a LastError
-		require.Nil(t, s.Description().LastError)
+		testCases := []struct {
+			name   string
+			err    error
+			conn   driver.Connection
+			result driver.ProcessErrorResult
+		}{
+			// One-off tests for errors that should have no effect.
+			{"nil error", nil, oldTVConn, driver.NoChange},
+			{"stale connection", errors.New("foo"), newStaleProcessErrorTestConn(), driver.NoChange},
+			{"non state change error", nonStateChangeError, oldTVConn, driver.NoChange},
+
+			// Tests for top-level (ok: 0) errors. We test a NotMaster error and a Shutdown error because the former
+			// only causes the server to be marked Unknown and the latter causes the pool to be cleared.
+			{"stale not master error", staleNotMasterError, newTVConn, driver.NoChange},
+			{"new not master error", newNotMasterError, oldTVConn, driver.ServerMarkedUnknown},
```

Comment:
No need to change, but I think these could all use `newTVConn`, right? SDAM would process these errors if the topology version is equal.

Reply:
Yeah, that should work too. All of the `new*` errors don't have a topology version, so they should never be considered as stale. I'm not going to change it because it works fine this way and I like that `newTVConn` has a distinct use.